packages/vite-plugin-cloudflare/src/plugin-config.ts

```diff
@@ -1,101 +1,189 @@
+
+export type WorkerConfig = Config & {
+	name: Defined<Config['name']>;
+	main: Defined<Config['main']>;
+};
+
+interface AssetsOnlyResult {
+	type: 'assets-only';
+	config: AssetsOnlyConfig;
 }
 
-export interface NormalizedPluginConfig {
-	workers: Record<
-		string,
-		{
-			entryPath: string;
-			wranglerConfigPath: string;
-			assetsBinding?: string;
-			workerOptions: SourcelessWorkerOptions & { name: string };
-		}
-	>;
-	entryWorkerName?: string;
-	assets: AssetConfig;
-	persistPath: string | false;
+interface WorkerResult {
+	type: 'worker';
+	config: WorkerConfig;
+}
+
+interface BasePluginConfig {
 	wranglerConfigPaths: Set<string>;
 }
 
-const DEFAULT_PERSIST_PATH = '.wrangler/state/v3';
+interface AssetsOnlyPluginConfig extends BasePluginConfig {
+	type: 'assets-only';
+	config: AssetsOnlyConfig;
+}
+
+interface WorkersPluginConfig extends BasePluginConfig {
+	type: 'workers';
+	workers: Record<string, WorkerConfig>;
+	entryWorkerEnvironmentName: string;
+	wranglerConfigPaths: Set<string>;
+}
+
+export type ResolvedPluginConfig = AssetsOnlyPluginConfig | WorkersPluginConfig;
+
+function getConfigResult(
+	configPath: string,
+	userConfig: vite.UserConfig,
+	wranglerConfigPaths: Set<string>,
+	isEntryWorker?: boolean,
+): AssetsOnlyResult | WorkerResult {
+	if (wranglerConfigPaths.has(configPath)) {
+		throw new Error(`Duplicate Wrangler config path found: ${configPath}`);
+	}
+
+	const wranglerConfig = readConfig(configPath, {
+		// We use the mode from the user config rather than the resolved config for now so that the mode has to be set explicitly. Otherwise, some things don't work as expected when `development` and `production` environments are not present.
+		env: userConfig.mode,
+	});
+
+	wranglerConfigPaths.add(configPath);
+
+	if (isEntryWorker && !wranglerConfig.main) {
+		invariant(
+			wranglerConfig.assets,
+			`No main or assets field provided in ${wranglerConfig.configPath}`,
+		);
+
+		return {
+			type: 'assets-only',
+			config: { ...wranglerConfig, assets: wranglerConfig.assets },
+		};
+	}
+
+	invariant(
+		wranglerConfig.main,
+		`No main field provided in ${wranglerConfig.configPath}`,
+	);
+
+	invariant(
+		wranglerConfig.name,
+		`No name field provided in ${wranglerConfig.configPath}`,
+	);
+
+	return {
+		type: 'worker',
+		config: {
+			...wranglerConfig,
+			name: wranglerConfig.name,
+			main: wranglerConfig.main,
+		},
+	};
+}
+
+// We can't rely on `readConfig` from Wrangler to find the config as it may be relative to a different root that's set by the user.
+function findWranglerConfig(root: string): string | undefined {
+	for (const extension of ['json', 'jsonc', 'toml']) {
+		const configPath = path.join(root, `wrangler.${extension}`);
+
+		if (fs.existsSync(configPath)) {
+			return configPath;
+		}
+	}
+}
+
+// Worker names can only contain alphanumeric characters and '_' whereas environment names can only contain alphanumeric characters and '$', '_'
```

Comment:
do we know why environment names don't allow `-`? maybe that's something that we can change upstream? (it is pretty minor in any case 😅)

Reply:
I assume there's probably a reason. Not sure why though.

Reply:
https://github.com/vitejs/vite/blob/81c31c16d10bff48c9965f7aedbf7efb7c7022c4/packages/vite/src/node/baseEnvironment.ts#L57-L58

feels like a pretty arbitrary reasoning... 🤷 

but hey, we can live with that 😅 🤷

Reply:
Hmm, I wouldn't have thought that `-` needs escaping in directory names. It would be nicer if they could be used so maybe it is worth looking into changing.